balancer.go

```diff
@@ -0,0 +1,256 @@
+			CipherSuites: []uint16{
+				tls.TLS_AES_128_GCM_SHA256,       // 2022 TLS v1.3 compliant
+				tls.TLS_CHACHA20_POLY1305_SHA256, // 2022 TLS v1.3 compliant
+				tls.TLS_AES_256_GCM_SHA384,       // 2022 TLS v1.3 compliant
+			},
+			MinVersion:       tls.VersionTLS13,
+			MaxVersion:       tls.VersionTLS13,
+			CurvePreferences: []tls.CurveID{tls.CurveP521, tls.CurveP384, tls.CurveP256},
+		}
+
+		scheduleListeners = append(scheduleListeners, schedule{port, config, identity})
+	}
+
+	// Schedule listeners one by one and fail if any of them fail
+	derCtx, derCancel := context.WithCancel(lb.runCtx)
+	errChan := make(chan error, len(scheduleListeners))
+	defer close(errChan)
+
+	wg := sync.WaitGroup{}
+	for _, params := range scheduleListeners {
+		go func(ctx context.Context, cancelAll context.CancelFunc, errChan chan error, toSchedule schedule) {
+
+			wg.Add(1)
+			// Don't forget to close all contexts
+			defer derCancel()
+			// Try to listen
+			listen, err := tls.Listen("tcp", fmt.Sprintf("localhost:%d", toSchedule.port), toSchedule.tls)
+			if err != nil {
+				errChan <- fmt.Errorf("failed to listen on port")
+				wg.Done()
+				return
+			}
+
+			// Spawn the coroutine to watch for the context break
+			go func(l net.Listener) {
+				// Here all required procedures were established for the listener thread
+				wg.Done()
+				// Await for context break
+				<-ctx.Done()
+				err := listen.Close()
+				lb.logger.Info(fmt.Sprintf("closing listener at port %d", toSchedule.port))
+				if err != nil {
+					lb.logger.Error(fmt.Sprintf("error closing listener at port %d", toSchedule.port))
+				}
+			}(listen)
+
+			lb.logger.Info(fmt.Sprintf("listening at port %d", toSchedule.port))
+
+			// Bind pool and forwarder
+			currentPool := toSchedule.pool
+			var forwarder *Forwarder
+
+			// Maintain incoming connections
+			for {
+
+				conn, err := listen.Accept()
+				if err != nil {
+					if strings.Contains(err.Error(), "closed network") {
+						// Graceful
+						lb.logger.Debug(fmt.Sprintf("listener closed, closed network for port: %d", toSchedule.port))
+						errChan <- nil
+					} else {
+						// Other kind of error
+						lb.logger.Error(fmt.Errorf("failed to accept connection, error: %w", err).Error())
+						errChan <- fmt.Errorf("failed to listen on port, error: %w", err)
+					}
+					delete(lb.forwarderMap, currentPool.Identity())
+					return
+				}
+
+				lb.logger.Debug(fmt.Sprintf("accepting request for port %d", toSchedule.port))
+
+				// Convert to TLS and proceed with the handshake
+				tlsConn := conn.(*tls.Conn)
+				err = tlsConn.Handshake()
+				if err != nil {
+					lb.logger.Error(fmt.Errorf("cannot complete handshake, %w", err).Error())
+					continue
+				}
+
+				// Verify certificate CN and find or create corresponding backend to dispatch
+				certs := tlsConn.ConnectionState().PeerCertificates
+				if len(certs) == 0 {
+					lb.logger.Error("failed to extract certificate")
+					continue
```

Comment:
Leaking resource - missing conn.Close call

Reply:
Thanks, I missed this case and missed this in tests it is too obvious that the connection should be closed there, and I believe this happened due to not paying enough attention to this particular area of execution rather than others. That's why we need as much eyes as possible :)